support/nfs/cacheio.c

```diff
@@ -258,3 +260,148 @@ cache_flush(int force)
 		}
 	}
 }
+
+/*
+ * Update the expiration of an existing entry in cache so that it gets evicted.
+ * cache is either 'nfsd.fh' or 'nfsd.export'
+ */
+static void
+cache_update_entry(const char *cache, const char *message)
+{
+	char path[MAXPATHLEN];
+	int fd;
+	ssize_t mesg_len = (ssize_t)strlen(message);
+
+	snprintf(path, sizeof(path), "/proc/net/rpc/%s/channel", cache);
+	fd = open(path, O_RDWR);
+	if (fd >= 0) {
+		if (write(fd, message, mesg_len) != mesg_len) {
+			xlog_warn("Writing '%s' to '%s' failed: errno %d (%s)",
+			message, path, errno, strerror(errno));
+		}
+		close(fd);
+	}
+}
+
+/*
+ * Check if there is an entry in cache for mountpoint.
+ * cache is either 'nfsd.fh' or 'nfsd.export'
+ */
+static bool
+cache_has_entry(const char *cache, const char *mountpoint)
+{
+	char path[MAXPATHLEN];
+	bool found = false;
+	FILE *fp;
+
+	snprintf(path, sizeof(path), "/proc/net/rpc/%s/content", cache);
+	fp = fopen(path, "r");
+	if (fp != NULL) {
+		char *linebuf = NULL;
+		size_t linelen = 0;
+
+		while(getline(&linebuf, &linelen, fp) != -1) {
+			if (linebuf[0] == '#')
+				continue;
+			if (strstr(linebuf, mountpoint) != NULL) {
+				found = true;
+				break;
+			}
+		}
+		free(linebuf);
+		(void) fclose(fp);
+	}
+
+	return found;
+}
+
+#define MILLISEC        1000
+#define NANOSEC         1000000000
+#define NSEC2MSEC(n)    ((n) / (NANOSEC / MILLISEC))
+
+static uint64_t
+gethrtime(void)
+{
+	struct timespec ts;
+	(void) clock_gettime(CLOCK_MONOTONIC, &ts);
+	return ((((uint64_t)ts.tv_sec) * NANOSEC) + ts.tv_nsec);
+}
+
+/*
+ * Check if mount is busy without causing an actual unmount.
+ */
+static bool
+can_umount(const char *path)
+{
+	struct statfs64 stfs;
+	int error = 0;
+
+	/*
+	 * We leverage the umount2 MNT_EXPIRE mechanism to let us know if
+	 * a mount is busy without causing an actual unmount.  With the
+	 * MNT_EXPIRE flag we will get back EBUSY or EAGAIN (mount is not
+	 * busy) but not cause an actual unmount.
+	 *
+	 * After checking the result we make a statfs call to clear the
+	 * MNT_EXPIRE state (i.e. the mnt_expiry_mark).
+	 */
+	if (umount2(path, MNT_EXPIRE) == -1)
+		error = errno;
+
+	/* Undo the mount point expired state by taking a temporary reference */
```

Comment:
Thanks for the updated comments, I think it's worth adding a comment about a potential race here where two processes can try to expire the same path. From our discussion this is fine and the rest of the code can handle that.